Update MASTG-KNOW-0036 and implement MASTG-TEST-0287 (#3706)
OWASP/mastg: knowledge/android/MASVS-STORAGE/MASTG-KNOW-0036.md

Reviewer: ```suggestion
The following example details storing data using `EncryptedSharedPreferences`, which encrypts the preference keys and values before writing them to disk:

```

Reviewer: Provided introduction to the code snippet

Reviewer: When reading https://developer.android.com/training/data-storage/shared-preferences, Android details: 
```
Caution: DataStore is a modern data storage solution that you should use instead of SharedPreferences. It builds on Kotlin coroutines and Flow, and overcomes many of the drawbacks of SharedPreferences.
```

Right now this PR focuses on testing the DataStore but did not create a KNOW file: https://github.com/OWASP/mastg/pull/3540.

I leave here just in case it's interesting to create a issue to track the KNOW file for the `DataStore`, although no encryption is provided by such mechanism right now.

Reviewer: Created https://github.com/OWASP/mastg/issues/3781 to add the `DataStore` KNOW file.

Reviewer: ```suggestion
The [`SharedPreferences`](https://developer.android.com/training/data-storage/shared-preferences "Shared Preferences") API is commonly used to permanently save small collections of key-value pairs in the app's sandbox storage.
```

Reviewer: Clarified where does the API store the data.

Reviewer: ```suggestion
Since Android 4.2 (API level 17), the `SharedPreferences` object can only be declared private and not world-readable, i.e., accessible to all apps. However, values stored in `SharedPreferences` are written to plain-text XML files in the app sandbox, so they can be extracted if the app data directory becomes accessible, for example through device compromise, backup extraction, or physical access to rooted or unlocked devices.
```

Reviewer: ```suggestion
Since Android 4.2 (API level 17) the `SharedPreferences` object can only be declared to be private (and not world-readable, i.e. accessible to all apps).

This is made by interacting with `SharedPreferences` by calling `getSharedPreferences` with the `Context.MODE_PRIVATE` flag (see ["Use SharedPreferences in private mode"](https://developer.android.com/privacy-and-security/security-best-practices#sharedpreferences).

When using such flag, the XML file containing the key-value data is stored with permissions that restrict access to the app's own Linux user ID. Under the normal Android app sandbox, other apps cannot read this file directly.

However, this flag does not provide encryption of the data, as the data is still written in plain-text in the XML file.
```

> `MODE_PRIVATE` makes the file only accessible by the calling app. See ["Use SharedPreferences in private mode"](https://developer.android.com/privacy-and-security/security-best-practices#sharedpreferences).

Reviewer: ```suggestion
```

Reviewer: Provided more context on the `MODE_PRIVATE` flag.

Author: @jacobocasado isn't this flag the only option in recent versions? It used to be an issue in the past if the dev used another one, right? Maybe we should point that out.

Reviewer: Yes! Indeed. I have merged this paragraph with the paragraph above, that also talks about this flag, specifying that now it's the only way to do so.

The annotation on the deprecated and unsafe modes is below, right after the example. Let me know if that is better!

> > Other insecure modes exist, such as `MODE_WORLD_READABLE` and `MODE_WORLD_WRITEABLE`, but they have been deprecated since Android 4.2 (API level 17) and [removed in Android 7.0 (API Level 24)](https://developer.android.com/reference/android/os/Build.VERSION_CODES#N). Therefore, only apps running on an older OS version (`android:minSdkVersion` less than 17) will be affected. Otherwise, Android will throw a [`SecurityException`](https://developer.android.com/reference/java/lang/SecurityException). If an app needs to share private files with other apps, it is best to use a [`FileProvider`](https://developer.android.com/reference/androidx/core/content/FileProvider) with the [`FLAG_GRANT_READ_URI_PERMISSION`](https://developer.android.com/reference/android/content/Intent#FLAG_GRANT_READ_URI_PERMISSION). See ["Sharing Files"](https://developer.android.com/training/secure-file-sharing) for more details.

Reviewer: The "If an app needs to share private files with other apps, it is best to use a [`FileProvider`]..."  is a recommendation, this might belong better in the MASTG-BEST file.

> You might also use [`EncryptedSharedPreferences`](https://developer.android.com/reference/androidx/security/crypto/EncryptedSharedPreferences) (see [source code](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:security/security-crypto/src/main/java/androidx/security/crypto/EncryptedSharedPreferences.java)), which is wrapper of `SharedPreferences` that automatically encrypts all data stored to the shared preferences using [Deterministic Authenticated Encryption with Associated Data (Deterministic AEAD) from Google's Tink Library](https://developers.google.com/tink/deterministic-encryption).

Reviewer: ```suggestion
[`EncryptedSharedPreferences`](https://developer.android.com/reference/androidx/security/crypto/EncryptedSharedPreferences) is a `SharedPreferences` wrapper from the Jetpack Security Crypto library. It encrypts preference keys with `AES256_SIV`, a deterministic Authenticated Encryption with Associated Data (AEAD) scheme, and preference values with `AES256_GCM`, an AEAD scheme, before writing them to disk. See the [`EncryptedSharedPreferences` source code](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:security/security-crypto/src/main/java/androidx/security/crypto/EncryptedSharedPreferences.java) for implementation details.
```

Reviewer: I have removed the "recommendation" on using `EncryptedSharedPreferences` over `SharedPreferences` as such recomendation is made in the BEST file we created (KNOW files should not recommend).

Also, the warning below might be in the KNOW file instead here, feel free to also remove it
